astropy/utils/data_info.py

```diff
@@ -346,7 +346,14 @@ def __set__(self, instance, value):
 
         if isinstance(value, DataInfo):
             info = instance.__dict__['info'] = self.__class__(bound=True)
-            for attr in info.attr_names - info.attrs_from_parent - info._attrs_no_copy:
+            attr_names = info.attr_names
+            if value.__class__ is self.__class__:
+                # For same class, attributes are guaranteed to be stored in
+                # _attrs, so speed matters up by not accessing defaults.
+                # Doing this before difference in for loop helps speed.
+                attr_names = attr_names & set(value._attrs)  # NOT in-place!
```

Comment:
This is the part that I started with, and now helps mostly for columns inside tables. It reduces slicing a simple quantity by 30%:
```
In [2]: q = [1, 2, 3] * u.s

In [3]: t = QTable({'q': q})

In [4]: %timeit t['q'][:2]
# current master
6.14 µs ± 11.8 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
# this PR
4.25 µs ± 21.4 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
```